### PR TITLE
feat: replace RMSNorm backward with persistent CuTile kernel

### PR DESCRIPTION
**Pull request by aghilann**
## Description

Replaces the old one-row-per-block RMSNorm backward kernel with a persistent grid-stride kernel that fuses `dw` accumulation into a compact `(grid × TILE_N)` partial-sum buffer instead of allocating an `M×N` temp buffer. I wrote this kernel and was able to get it to exceed the performance of Liger's Triton Kernels and get quite close to the performance of Quacks CuteDSL Kernel (which I'm somewhat assuming is near peak-performance)

**Changes:**
- **Backward kernel**: New persistent `_rms_bwd` with grid-stride loop and fused `dw` accumulation
- **Forward kernels**: Unchanged (gather + static persistent), except persistent now also stores `rstd` so backward works from both modes
- **Reference fix**: `rms_norm_backward_torch` now casts to fp32 upfront, matching kernel precision
- Removed unused aliases and imports

**Backward kernel throughput (GB/s)** — standalone, M=4096

| N | Old Version | **New Version** | PyTorch | Old→New | vs PyTorch |
|---:|---:|---:|---:|---:|---:|
| **bf16** |
| 1024 | 1,534 | **3,020** | 482 | 2.0x | 6.3x |
| 2048 | 2,274 | **6,709** | 549 | 2.9x | 12.2x |
| 4096 | 2,823 | **6,100** | 538 | 2.2x | 11.3x |
| 8192 | 3,652 | **8,008** | 552 | 2.2x | 14.5x |
| 16384 | 3,762 | **4,135** | 573 | 1.1x | 7.2x |
| **fp16** |
| 1024 | 1,561 | **3,197** | 483 | 2.0x | 6.6x |
| 2048 | 2,368 | **4,454** | 553 | 1.9x | 8.1x |
| 4096 | 2,933 | **6,131** | 540 | 2.1x | 11.4x |
| 8192 | 3,578 | **7,987** | 553 | 2.2x | 14.4x |
| 16384 | 4,020 | **4,332** | 574 | 1.1x | 7.5x |
| **fp32** |
| 1024 | 2,268 | **4,940** | 955 | 2.2x | 5.2x |
| 2048 | 2,923 | **7,357** | 1,099 | 2.5x | 6.7x |
| 4096 | 3,901 | **7,140** | 1,030 | 1.8x | 6.9x |
| 8192 | 3,667 | **9,298** | 1,051 | 2.5x | 8.8x |
| 16384 | 2,634 | **6,943** | 1,090 | 2.6x | 6.4x |

## CI Configuration

```yaml
config:
  build: true
  # valid options are "ops" and "benchmark"
  test: ["ops", "benchmark"]
```

## Checklist
- [x] Code formatted and imports sorted via repo specifications (`./format.sh`)
- [ ] Documentation updated (if needed)
- [ ] CI configuration reviewed

### CODE REVIEW

**Comment by aghilann** [comment]
Could I get a review when you get a chance @hannahli-nv  !

**Comment by aghilann** [comment]
@hannahli-nv Sorry to double ping, wanted to get this reviewed soon since I have a few more on the way with good performance :)

**Comment by hannahli-nv** [comment]
The existing [Test_RMSNormBackward](https://github.com/NVIDIA/TileGym/blob/main/tests/ops/experimental/test_rmsnorm_backward.py#L15) tests the standalone `rms_norm_backward()` with [pytorch-computed rstd](https://github.com/NVIDIA/TileGym/blob/main/tests/ops/experimental/test_rmsnorm_backward.py#L57). It does not exercise the `RMSNorm.backward()` autograd path, nor does it verify that the forward kernels store rstd correctly. Could you please add a `Test_RMSNormAutogradBackward` class in [test_rmsnorm_backward.py](https://github.com/NVIDIA/TileGym/blob/main/tests/ops/experimental/test_rmsnorm_backward.py) that runs rms_norm() forward with `requires_grad=True` inputs and validates gradients via autograd?

**Comment by hannahli-nv** [comment]
[bench_rmsnorm_backward.py](https://github.com/NVIDIA/TileGym/blob/main/tests/benchmark/experimental/bench_rmsnorm_backward.py) still assumes the old M x N temp buffer instead of the partial-sum buffer. Could you please update it to reflect the new partial-sum buffer size, as this will ensure the correct GBps improvement.

File: src/tilegym/ops/cutile/rms_norm.py [hunk 1]
```diff
@@ -265,6 +145,104 @@ def rms_norm_kernel_static_persistent(
         )
 
 
+@ct.kernel(occupancy=1)
```
Comment (hannahli-nv) [inline src/tilegym/ops/cutile/rms_norm.py]: Per [CONTRIBUTING.md](https://github.com/NVIDIA/TileGym/blob/main/CONTRIBUTING.md#situation-a-add-a-new-kernel-or-a-new-op), new kernels should carry the `@experimental_kernel `decorator. Please add the `@experimental_kernel `decorator to the `new _rms_bwd` kernel as was done with [the content you deleted](https://github.com/NVIDIA/TileGym/blob/main/src/tilegym/ops/cutile/rms_norm.py#L17). We will remove the experimental marker once its functional correctness and performance have been fully validated. Thank you for your understanding.

**Comment by hannahli-nv** [comment]
> @hannahli-nv Sorry to double ping, wanted to get this reviewed soon since I have a few more on the way with good performance :)

Sorry for the delayed response, as I was out of the office over the past few days. I have added a few comments. Thanks again for your excellent optimization!

**Comment by aghilann** [comment]
Hey @hannahli-nv, I’ve made the requested changes. Thank you for the thorough review. I  saw the updated CONTRIBUTING.MD, I'm excited for the roadmap! Also, I was running this on a consumer Blackwell chip since I don't have a B200 with CUDA 13.1 on hand, so the benchmarks in the PR description might need to be updated due to the update to the partial-buffer size calculation change.

File: src/tilegym/ops/cutile/rms_norm.py [hunk 2]
```diff
@@ -374,31 +352,53 @@ def ceil_div(a, b):
...
         x, weight, rstd = ctx.saved_tensors
+        shape = x.shape
+        N = shape[-1]
+        x2 = x.reshape(-1, N)
+        M = x2.shape[0]
+        dy2 = dy.reshape(-1, N)
+        if not dy2.is_contiguous():
+            dy2 = dy2.contiguous()
+
+        cfg = _bwd_cfg.get((M, N))
```
[the remark this answers is not in the document]
Comment (aghilann) [reply src/tilegym/ops/cutile/rms_norm.py]: My apologies, I should have caught this earlier.

**Comment by hannahli-nv** [comment]
/ok to test https://github.com/NVIDIA/TileGym/commit/abb1556542cd7f564544fa0f8eba67ac86b7a23b

**Comment by hannahli-nv** [comment]
/ok to test https://github.com/NVIDIA/TileGym/commit/32c2a51adb40ad7ff6c1fa86d58a28c31ec17dc7

**Review by hannahli-nv — APPROVED** [review]
LGTM, thx!
